message_ix/tests/test_feature_vintage_and_active_years.py

```diff
@@ -0,0 +1,344 @@
+    obs = scen.vintage_and_active_years(
+        ya_args=("foo", "bar", years[0]), in_horizon=True
+    )
+    pdt.assert_frame_equal(
+        pd.DataFrame(columns=["year_vtg", "year_act"]), obs, check_dtype=False
+    )
+
+    # Exception is raised for incorrect arguments
+    with pytest.raises(
+        ValueError, match="At least 2 arguments are required if using `ya_args`"
+    ):
+        scen.vintage_and_active_years(ya_args=("foo"))
+
+
+def test_vintage_and_active_years2(test_mp):
+    scen = Scenario(test_mp, **SCENARIO["dantzig"], version="new")
+
+    # Add years of differing time length
+    years = [2000, 2005, 2010, 2015, 2020, 2030]
+    scen.add_horizon(year=years, firstmodelyear=2010)
+
+    # Check if default function call is valid
+    obs = scen.vintage_and_active_years()
+    exp = pd.DataFrame(
+        {
+            "year_vtg": (
+                2000,
+                2000,
+                2000,
+                2000,
+                2005,
+                2005,
+                2005,
+                2005,
+                2010,
+                2010,
+                2010,
+                2010,
+                2015,
+                2015,
+                2015,
+                2020,
+                2020,
+                2030,
+            ),
+            "year_act": (
+                2010,
+                2015,
+                2020,
+                2030,
+                2010,
+                2015,
+                2020,
+                2030,
+                2010,
+                2015,
+                2020,
+                2030,
+                2015,
+                2020,
+                2030,
+                2020,
+                2030,
+                2030,
+            ),
+        }
+    )
+    pdt.assert_frame_equal(exp, obs, check_like=True)  # ignore col order
+
+    # Add a technology, its lifetime, and period durations
+    scen.add_set("node", "foo")
+    scen.add_set("technology", "bar")
+    scen.add_par(
+        "duration_period",
+        pd.DataFrame({"unit": "???", "value": [5, 5, 5, 5, 5, 10], "year": years}),
+    )
+    scen.add_par(
+        "technical_lifetime",
+        pd.DataFrame(
+            {
+                "node_loc": "foo",
+                "technology": "bar",
+                "unit": "???",
+                "value": 20,
+                "year_vtg": years,
+            }
+        ),
+    )
+
+    # Check standard functionality with different period duration lengths
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar", "2010"))
+    exp = pd.DataFrame(
+        {"year_vtg": (2010, 2010, 2010, 2010), "year_act": (2010, 2015, 2020, 2030)}
+    )
+    pdt.assert_frame_equal(exp, obs, check_like=True)  # ignore col order
+
+    # Check if no vintge-year is passed, that all values corresponding
+    # to technical lifetime are passed if the active years >= 2010
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar"))
+    exp = pd.DataFrame(
```

Comment:
@OFR-IIASA can you please help me understand this expected value?
- `technical_lifetime` is 20 for all periods.
- The pair yv=2000, ya=2020 and yv=2000, ya=2030 are excluded. I guess this is because 2020 - 2000 ≥ 20.
- However, the pair yv=2010, ya=2030 are *in*cluded, even though 2030 - 2010 ≥ 20.

What is the difference?

Reply:
Thanks, understood. This accords with our documentation.